### PR TITLE
Dashboard: Support additional variable format options (singlequote, doublequote, sqlstring)

### PR DESCRIPTION
**Pull request by xiaobeiyang**
**What this PR does / why we need it**:
This PR adds 3 formatting options of variables, mainly for SQL. 
New formatting list:
    singlequote: Quote variable value(s) with single quote, escape single quotes in original value with backslashes.
    doublequote: Quote variable value(s) with double quote, escape double quotes in original value with backslashes.
    sqlstring: Quote variable value(s) with single quote, escape `'` in original value by `''`.

When writing SQL statements, we usually need  this kind of statement “select * from table1 where name in (${var})”, but Grafana doesn’t provides formatting option to quote string of SQL.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by aknuds1** [comment]
Could you please fill in the section 'What this PR does / why we need it' in the PR template?

**Comment by xiaobeiyang** [comment]
@aknuds1  Thanks for your quick response, PR template has been update.

**Comment by torkelo** [comment]
There should be custom format functions in the Postgres and MySQL datasources I’m that does this I think

**Comment by xiaobeiyang** [comment]
Thanks @torkelo . It is not easy to decide which variable should be quoted/escaped in datasources plugins. For example: `select $var1, $var2 from $tableVar where col1=$colVar1 AND col2 in $colVar2` . 
I'm writing plugin for other database service instead of Postgres and MySQL. It will be great if Grafana platform provides these formats out of the box, then datasource plugins could have more common behaviors both in developing & using.

File: docs/sources/reference/templating.md [hunk 1]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
 Interpolation result: 'foo%28%29bar%20BAZ%2Ctest2'
 ```
 
+### Singlequote
+Formats single and multi valued variables for use in URL parameters.
```
Comment (oddlittlebird) [inline docs/sources/reference/templating.md]:
```suggestion
Formats single- and multi-valued variables for use in URL parameters.
```

File: docs/sources/reference/templating.md [hunk 2]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+
+```bash
+servers = ['test1', 'test2']
+String to interpolate: '${servers:singlequote}'
+Interpolation result: "'test1','test2'"
+```
+
+### Doublequote
+Formats single and multi valued variables for use in URL parameters.
```
Comment (oddlittlebird) [inline docs/sources/reference/templating.md]:
```suggestion
Formats single- and multi-valued variables for use in URL parameters.
```

File: docs/sources/reference/templating.md [hunk 3]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+
+```bash
+servers = ['test1', 'test2']
+String to interpolate: '${servers:doublequote}'
+Interpolation result: '"test1","test2"'
+```
+
+### Sqlstring
+Formats single and multi valued variables for use in URL parameters.
```
Comment (oddlittlebird) [inline docs/sources/reference/templating.md]:
```suggestion
Formats single- and multi-valued variables for use in URL parameters.
```

**Review by oddlittlebird — CHANGES_REQUESTED** [review]

File: docs/sources/reference/templating.md [hunk 4]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+Formats single- and multi-valued variables for use in URL parameters.
+
+```bash
+servers = ['test1', 'test2']
+String to interpolate: '${servers:doublequote}'
+Interpolation result: '"test1","test2"'
+```
+
+### Sqlstring
+Formats single- and multi-valued variables for use in URL parameters.
```
Comment (torkelo) [inline docs/sources/reference/templating.md]: for use in URL parameters? should that not be SQL parameters?

File: docs/sources/reference/templating.md [hunk 5]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+
+```bash
+servers = ["test'1", "test2"]
+String to interpolate: '${servers:sqlstring}'
+Interpolation result: "'test''1','test2'"
```
Comment (torkelo) [inline docs/sources/reference/templating.md]: how is this different from singlequote ?

File: docs/sources/reference/templating.md (hunk 4 above)
Comment (xiaobeiyang) [reply docs/sources/reference/templating.md]: Sorry, I will update the doc.

File: docs/sources/reference/templating.md (hunk 5 above)
Comment (xiaobeiyang) [reply docs/sources/reference/templating.md]: Singlequote just quote value with `'`, while sqlstring escape `'` by `''` first, and then quote the value. Sqlstring results is safe in SQL statements. The 2 formats are designed for different scenarios.

**Review by ryantxu — DISMISSED** [review]
It is nice to have examples in the docs... but I would be waaaay happier with examples in tests :)

**Comment by xiaobeiyang** [comment]
@ryantxu  Please help review. Thanks.

**Review by xiaobeiyang — COMMENTED** [review]
updated

File: docs/sources/reference/templating.md [hunk 6]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+```
+
+### Doublequote
+Formats single- and multi-valued variables into a comma-separated string, quotes each value with `"`.
+
+```bash
+servers = ['test1', 'test2']
+String to interpolate: '${servers:doublequote}'
+Interpolation result: '"test1","test2"'
+```
```
Comment (torkelo) [inline docs/sources/reference/templating.md]: without a way to escape the values (single or double quotes), won't that limit the usefulness of these?

Comment (xiaobeiyang) [reply docs/sources/reference/templating.md]: yes, I added the escape logic.

**Comment by xiaobeiyang** [comment]
Hi @torkelo , is there any other problem in the PR or some process? We expect this RP will be merged and start use it in my dashboards.  Please help review, I will fix if there is any problem.
Thanks.

File: docs/sources/reference/templating.md [hunk 7]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
...
+Formats single- and multi-valued variables into a comma-separated string, escapes single quote in value and quotes each value with `'`.
+
+```bash
+servers = ['test1', 'test2']
+String to interpolate: '${servers:singlequote}'
+Interpolation result: "'test1','test2'"
+```
+
+### Doublequote
+Formats single- and multi-valued variables into a comma-separated string, escapes double quote in value and quotes each value with `"`.
```
Comment (torkelo) [inline docs/sources/reference/templating.md]: Can we add how this is escaped

File: docs/sources/reference/templating.md [hunk 8]
```diff
@@ -114,6 +114,33 @@ String to interpolate: '${servers:percentencode}'
 Interpolation result: 'foo%28%29bar%20BAZ%2Ctest2'
 ```
 
+### Singlequote
+Formats single- and multi-valued variables into a comma-separated string, escapes single quote in value and quotes each value with `'`.
```
Comment (torkelo) [inline docs/sources/reference/templating.md]: Can you add how the value is escaped

**Review by torkelo — APPROVED** [review]
Really sorry for slow feedback, this looks great now. Just a small detail missing in the docs

File: docs/sources/reference/templating.md (hunk 8 above)
Comment (xiaobeiyang) [reply docs/sources/reference/templating.md]: updated

**Comment by xiaobeiyang** [comment]
@torkelo  @ryantxu @oddlittlebird  Please help review, many thanks.

**Review by oddlittlebird — APPROVED** [review]

**Comment by oddlittlebird** [comment]
LGTM!

**Review by xiaobeiyang — COMMENTED** [review]
please help review

File: public/app/features/templating/specs/template_srv.test.ts [hunk 9]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
       const result = _templateSrv.formatValue('Gi3/14', 'regex');
       expect(result).toBe('Gi3\\/14');
     });
+
+    it('single value and singlequote should render quoted string, value is quoted by single quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('single value and singlequote format should render string with value enclosed in single quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 10]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue('test', 'singlequote');
+      expect(result).toBe("'test'");
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
      expect(result).toBe(`'test'`);
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 11]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+    });
+
+    it('multi value and singlequote should render quoted string, each value is quoted by single quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('multi value and singlequote format should render string with values enclosed in single quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 12]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue(['test', "test'2"], 'singlequote');
+      expect(result).toBe("'test','test\\'2'");
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
      expect(result).toBe(`'test','test\\'2'`);
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 13]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+    });
+
+    it('single value and doublequote should render quoted string, value is quoted by double quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('single value and doublequote format should render string with value enclosed in double quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 14]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue('test', 'doublequote');
+      expect(result).toBe('"test"');
+    });
+
+    it('multi value and doublequote should render quoted string, each value is quoted by double quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('multi value and doublequote format should render string with values enclosed in double quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 15]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue(['test', 'test"2'], 'doublequote');
+      expect(result).toBe('"test","test\\"2"');
+    });
+
+    it('single value and sqlstring should render sql string, value is escaped and quoted by single quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('single value and sqlstring format should render string with value enclosed in single quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 16]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue("test'value", 'sqlstring');
+      expect(result).toBe("'test''value'");
+    });
+
+    it('multi value and sqlstring should render sql string, each value is escaped and quoted by single quote', () => {
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
    it('multi value and sqlstring format should render string with values enclosed in single quotes', () => {
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 17]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+      const result = _templateSrv.formatValue(['test', "test'value2"], 'sqlstring');
+      expect(result).toBe("'test','test''value2'");
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
      expect(result).toBe(`'test','test''value2'`);
```

File: public/app/features/templating/specs/template_srv.test.ts [hunk 18]
```diff
@@ -330,6 +330,36 @@ describe('templateSrv', () => {
...
+    });
+
+    it('multi value and doublequote should render quoted string, each value is quoted by double quote', () => {
+      const result = _templateSrv.formatValue(['test', 'test"2'], 'doublequote');
+      expect(result).toBe('"test","test\\"2"');
+    });
+
+    it('single value and sqlstring should render sql string, value is escaped and quoted by single quote', () => {
+      const result = _templateSrv.formatValue("test'value", 'sqlstring');
+      expect(result).toBe("'test''value'");
```
Comment (aknuds1) [inline public/app/features/templating/specs/template_srv.test.ts]:
```suggestion
      expect(result).toBe(`'test''value'`);
```

File: public/app/features/templating/template_srv.ts [hunk 19]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
         }
         return this.encodeURIComponentStrict(value);
       }
+      case 'singlequote': {
+        // escape single quotes with slash
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        // escape single quotes with backslash
```

File: public/app/features/templating/template_srv.ts [hunk 20]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        const regExp = new RegExp("'", 'g');
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        const regExp = new RegExp(`'`, 'g');
```

File: public/app/features/templating/template_srv.ts [hunk 21]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        if (_.isArray(value)) {
+          return _.map(value, v => `'${_.replace(v, regExp, "\\'")}'`).join(',');
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
          return _.map(value, v => `'${_.replace(v, regExp, `\\'`)}'`).join(',');
```

File: public/app/features/templating/template_srv.ts [hunk 22]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        }
+        return `'${_.replace(value, regExp, "\\'")}'`;
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        return `'${_.replace(value, regExp, `\\'`)}'`;
```

File: public/app/features/templating/template_srv.ts [hunk 23]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+      }
+      case 'doublequote': {
+        // escape double quotes with slash
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        // escape double quotes with backslash
```

File: public/app/features/templating/template_srv.ts [hunk 24]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        const regExp = new RegExp('"', 'g');
+        if (_.isArray(value)) {
+          return _.map(value, v => `"${_.replace(v, regExp, '\\"')}"`).join(',');
+        }
+        return `"${_.replace(value, regExp, '\\"')}"`;
+      }
+      case 'sqlstring': {
+        // escape single quotes in sql string
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        // escape single quotes by pairing them
```

File: public/app/features/templating/template_srv.ts [hunk 25]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        const regExp = new RegExp("'", 'g');
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        const regExp = new RegExp(`'`, 'g');
```

File: public/app/features/templating/template_srv.ts [hunk 26]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        if (_.isArray(value)) {
+          return _.map(value, v => `'${_.replace(v, regExp, "''")}'`).join(',');
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
          return _.map(value, v => `'${_.replace(v, regExp, `''`)}'`).join(',');
```

File: public/app/features/templating/template_srv.ts [hunk 27]
```diff
@@ -182,6 +182,30 @@ export class TemplateSrv {
...
+        }
+        return `'${_.replace(value, regExp, "''")}'`;
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        return `'${_.replace(value, regExp, `''`)}'`;
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Please see my suggestions.

**Comment by aknuds1** [comment]
Hang on, going to adjust my suggestions as I hear that double quotes are allowed (I'm used to them not being allowed).

**Comment by aknuds1** [comment]
OK, edited my suggestions :)

**Comment by xiaobeiyang** [comment]
Thanks @aknuds1 . Please help review again. 
The CI `grafana-docker-ubuntu-pr` step failed due to Ubuntu repository issue. I don't have rerun privilege. It looks like grafana-docker-ubuntu-pr fail didn't cause by the new commits.

File: public/app/features/templating/template_srv.ts [hunk 28]
```diff
@@ -183,24 +183,24 @@ export class TemplateSrv {
...
         const regExp = new RegExp('"', 'g');
         if (_.isArray(value)) {
           return _.map(value, v => `"${_.replace(v, regExp, '\\"')}"`).join(',');
         }
         return `"${_.replace(value, regExp, '\\"')}"`;
       }
       case 'sqlstring': {
-        // escape single quotes in sql string
-        const regExp = new RegExp("'", 'g');
+        // escape single quotes in by pairing them
```
Comment (aknuds1) [inline public/app/features/templating/template_srv.ts]:
```suggestion
        // escape single quotes by pairing them
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Thanks for applying my suggestions, I just saw a remaining typo.

**Comment by aknuds1** [comment]
Can you please merge this branch with Grafana master, in order to fix a CI failure?

File: public/app/features/templating/template_srv.ts (hunk 28 above)
Comment (xiaobeiyang) [reply public/app/features/templating/template_srv.ts]: updated

**Review by aknuds1 — APPROVED** [review]
LGTM

**Comment by aknuds1** [comment]
@torkelo @marefr This is good to merge right? Shall we add a milestone?

**Comment by aknuds1** [comment]
I'm just pushing a merge with latest Grafana master to your branch, in order to fix CI.

**Review by torkelo — APPROVED** [review]

**Comment by aknuds1** [comment]
Thanks very much for contributing to Grafana!

**Pull request merged**